### PR TITLE
Internal link

### PR DESCRIPTION
**Pull request by lecunm**

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "test": "ng test",
     "test-no-watch": "ng test --watch=false --progress=false",
     "lint": "ng lint",
-    "pree2e": "webdriver-manager update --gecko=false",
+    "pree2e": "webdriver-manager update --versions.chrome 78.0.3904.70 --gecko=false ",
```
Comment (loriepisicchio) [inline frontend/package.json]: Remove this config as it was just to fix the tests

File: frontend/package.json [hunk 2]
```diff
@@ -38,10 +38,12 @@
     "ngx-markdown": "8.2.1",
     "ngx-moment": "3.4.0",
     "ngx-numeral": "0.0.1-beta.0",
+    "protractor": "^5.4.2",
```
Comment (loriepisicchio) [inline frontend/package.json]: Versions should be fix

File: frontend/package.json [hunk 3]
```diff
@@ -38,10 +38,12 @@
...
     "rxjs": "6.5.3",
     "rxjs-compat": "6.5.3",
     "ts-option": "2.0.0",
     "tslib": "1.10.0",
+    "webdriver": "^5.15.4",
```
Comment (loriepisicchio) [inline frontend/package.json]: Same here: version should be fix

File: frontend/src/app/internal-link.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  constructor(private activatedRoute: ActivatedRoute) {
+  }
+
+
+  transform(value: string): string {
+    const linkRegexString = '(href=)["\'](thegardener:\\/\\/)(\\w*)?\\/?(\\w*)?;?path=(.*?)["\']';
+    const linkRegex = new RegExp(linkRegexString, 'g');
+    const hierarchy = this.activatedRoute.parent.snapshot.params.name;
+
+    function replacer(p1: string, p2: string, p3: string, p4: string, p5: string, p6: string) {
```
Comment (loriepisicchio) [inline frontend/src/app/internal-link.pipe.ts]: parameter names must be more explicit

File: frontend/src/app/output/page-content/page-content.component.ts [hunk 5]
```diff
@@ -12,16 +12,20 @@ import {NotificationService} from '../../_services/notification.service';
...
               private pageService: PageService,
-              private notificationService: NotificationService) {
+              private notificationService: NotificationService,
+              private router: Router,
+              private ngZone: NgZone) {
   }
 
   ngOnInit() {
+    // @ts-ignore
+    window.navigateTo = navigateTofun(this.router, this.ngZone);
```
Comment (loriepisicchio) [inline frontend/src/app/output/page-content/page-content.component.ts]: Rename `navigatTofun`

File: frontend/src/app/output/page-content/page-content.component.ts [hunk 6]
```diff
@@ -63,4 +67,9 @@ export class PageContentComponent implements OnInit, OnDestroy  {
   getScenario(part: PagePart) {
     return (part.data as ScenarioPart).scenarios;
   }
+
 }
+
+const navigateTofun = (router: Router, ngZone: NgZone) => (path: string) => {
+  ngZone.run(() => router.navigateByUrl(path)).then();
```
Comment (loriepisicchio) [inline frontend/src/app/output/page-content/page-content.component.ts]: We could at least log the error in `then` if Promise fails

File: frontend/src/app/removehtmlsanitizer.pipe.spec.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import {RemoveHtmlSanitizerPipe} from './removehtmlsanitizer.pipe';
+import {DomSanitizer} from '@angular/platform-browser';
+import {TestBed} from '@angular/core/testing';
+
+describe('RemovehtmlsanitizerPipe', () => {
+  it('create an instance', () => {
+    const sanitizer: DomSanitizer = TestBed.get(DomSanitizer);
+    const pipe = new RemoveHtmlSanitizerPipe(sanitizer);
+    expect(pipe).toBeTruthy();
```
Comment (loriepisicchio) [inline frontend/src/app/removehtmlsanitizer.pipe.spec.ts]: Is it possible to add a quick test with an HTML that was sanitized before ?

File: .azure-pipelines/frontend.yml [hunk 8]
```diff
@@ -4,66 +4,65 @@ pool:
...
+      cd ./node_modules/protractor && npm i webdriver-manager@latest && cd ../..
+      npm run pree2e
+    workingDirectory: frontend
+    displayName: 'npm install'
 
-- script: |
-    export CI=true
-    ./node_modules/tslint/bin/tslint -t json -o report.json --project . --force
-    ng test --watch=false --progress=false --code-coverage
-    ng e2e
```
Comment (notflorian) [inline .azure-pipelines/frontend.yml]: Why removed End To End Test? Please add it back.

File: documentation/features/navigation.md [hunk 9]
```diff
@@ -1,2 +1,8 @@
 
-TODO...
\ No newline at end of file
+TODO...
```
Comment (notflorian) [inline documentation/features/navigation.md]: Remove `TODO...`

File: .azure-pipelines/frontend.yml (hunk 8 above)
Comment (lecunm) [reply .azure-pipelines/frontend.yml]: we removed it because it was causing a bug with webdriver and we don't have End To End tests

Comment (notflorian) [reply .azure-pipelines/frontend.yml]: We should have end to end tests, and not sweep under the rug the issue with webdriver. 
E2E tests are very useful to detect issue with angular templates compilation, dependencies injection, AOT... Could you please add a simple E2E test to test that the application load and is properly displayed?

Comment (loriepisicchio) [reply .azure-pipelines/frontend.yml]: So far we have no e2e tests, and when we'll take time to write some, I think we'd rather use Cypress. So IMHO no need to waste our time trying to fix this protractor / webdriver issue.

Comment (notflorian) [reply .azure-pipelines/frontend.yml]: Ok, I'll close this comment and I will create another issue to add proper e2e tests.

**Comment by GeReinhart** [comment]
Done :
- put back commit from master to be able to merge. 
- put the documentation on the usage of this feature under documentation/guides/write.md 
- set up the dev env to test it : http://dc1-thegardener-dev-srv-01.dev.dc1.kelkoo.net:9000/app/documentation/navigate/_tools;path=theGardener%3Efeature_70-internal-link%3E_guides_/write

It doesn't seem to write well. 

Todo:
- make it work on the DEV platform to be able to validate it
- take in account remarks above

File: frontend/package.json [hunk 10]
```diff
@@ -9,23 +9,23 @@
...
     "test-no-watch": "ng test --watch=false --progress=false",
     "lint": "ng lint",
-    "pree2e": "webdriver-manager update --gecko=false",
+    "pree2e": "webdriver-manager update --gecko=false ",
     "e2e": "ng e2e --webdriver-update=false"
   },
   "private": true,
   "dependencies": {
-    "@angular/animations": "8.2.13",
+    "@angular/animations": "8.2.11",
```
Comment (loriepisicchio) [inline frontend/package.json]: why downgrading ? Or maybe you just miss a rebase with master

**Pull request merged**